### PR TITLE
Update EndBlock parameters

### PR DESCRIPTION
**Pull request by ethanfrey**
* Update abci dependencies
* Modify references from Diffs to Changes
* Fixes issues #924
* Partial fix for #952 
  * Updates consensus parameter upon EndBlock
  * Add Hash for ConsensusParams to block headers
  * State maintains LastConsensusHash to verify blocks

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -251,24 +251,68 @@ func (s *State) SetBlockAndValidators(header *types.Header, blockPartsHeader typ
...
 		header.NumTxs,
 		types.BlockID{header.Hash(), blockPartsHeader},
 		header.Time,
-		prevValSet, nextValSet)
+		prevValSet, nextValSet,
+		nextParams)
+
+}
+
+// applyChanges returns a new param set, overriding any
```
Comment (odeke-em) [inline state/state.go]: How about?
```go
// applyChanges returns new ConsensusParams
// whose fields are set to any non-zero fields of c.
// If c is nil, it returns p unmodified, as it was passed in.
```

Comment (ethanfrey) [reply state/state.go]: done!

Comment (odeke-em) [reply state/state.go]: Thank you @ethanfrey!

**Comment by ethanfrey** [comment]
Need help debugging this.

Make vagrant_test works.
CI times out. 

Whatever

**Comment by melekes** [comment]
```
E[12-14|23:06:21.115] Error signing vote                           module=consensus height=14 round=0 vote="Vote{0:DFCCC3603B2C 14/00/1(Prevote) BE4288EAF2D4 {<nil>} @ 2017-
12-14T23:06:21.115Z}" err="Error signing vote: Step regression"
```

File: types/block.go [hunk 2]
```diff
@@ -91,6 +92,9 @@ func (b *Block) ValidateBasic(chainID string, lastBlockHeight int64,
 	if !bytes.Equal(b.AppHash, appHash) {
 		return errors.New(cmn.Fmt("Wrong Block.Header.AppHash.  Expected %X, got %v", appHash, b.AppHash))
 	}
+	if !bytes.Equal(b.ConsensusHash, consensusHash) {
+		return errors.New(cmn.Fmt("Wrong Block.Header.ConsensusHash.  Expected %X, got %v", consensusHash, b.ConsensusHash))
```
Comment (odeke-em) [inline types/block.go]: cmn.Fmt is deprecated, in the future I think just `fmt.Sprintf` would suffice ;)

File: types/params_test.go [hunk 3]
```diff
@@ -38,3 +40,49 @@ func TestConsensusParamsValidation(t *testing.T) {
...
+		makeParams(6, 5, 4, 3, 2, 1),
+	}
+
+	hashes := make([][]byte, len(params))
+	for i := range params {
+		hashes[i] = params[i].Hash()
+	}
+
+	// make sure there are no duplicates...
+	// sort, then check in order for matches
```
Comment (odeke-em) [inline types/params_test.go]: Neat!

File: state/state.go [hunk 4]
```diff
@@ -251,24 +254,76 @@ func (s *State) SetBlockAndValidators(header *types.Header, blockPartsHeader typ
...
+
+	if c == nil {
+		return p
+	}
+	res := p
+	// we must defensively consider any structs may be nil
+	if c.BlockSize != nil {
+
+		if c.BlockSize.MaxBytes != 0 {
+			res.BlockSize.MaxBytes = int(c.BlockSize.MaxBytes)
```
Comment (melekes) [inline state/state.go]: should we check this is greater than `0`?

**Comment by ebuchman** [comment]
Thanks. I've got it from here. Let me merge the other PR in and then we can do final review

File: circle.yml [hunk 5]
```diff
@@ -25,7 +25,6 @@ dependencies:
 
 test:
   override:
-    - cd "$PROJECT_PATH" && make tools && make get_vendor_deps && make metalinter_test
```
[the remark this answers is not in the document]
Comment (ebuchman) [reply circle.yml]: It was failing with out of memory errors. I don't have time for that right now.

**Pull request merged**